pkg/agent/route/route.go

```diff
@@ -0,0 +1,304 @@
+		ServiceRtTable.Name = AntreaServiceTable
+	}
+
+	if ServiceRtTable.IsMainTable() {
+		_ = c.RemoveServiceRouting()
+		return nil
+	}
+	f, err := os.OpenFile(routeTableConfigPath, os.O_RDWR|os.O_APPEND, 0)
+	if err != nil {
+		klog.Fatalf("Unable to create service route table(open):  %v", err)
+	}
+	defer func() { _ = f.Close() }()
+
+	oldTablesRaw := make([]byte, 1024)
+	bLen, err := f.Read(oldTablesRaw)
+	if err != nil {
+		klog.Fatalf("Unable to create service route table(read): %v", err)
+	}
+	oldTables := string(oldTablesRaw[:bLen])
+	newTable := fmt.Sprintf("%d %s", ServiceRtTable.Idx, ServiceRtTable.Name)
+
+	if strings.Index(oldTables, newTable) == -1 {
+		if _, err := f.WriteString(newTable); err != nil {
+			klog.Fatalf("Failed to add antrea service route table: %v", err)
+		}
+	}
+
+	gwConfig := c.nodeConfig.GatewayConfig
+	if gwConfig != nil && c.nodeConfig.PodCIDR != nil {
+		// add local podCIDR if applicable to service rt table
+		route := &netlink.Route{
+			LinkIndex: util.GetNetLink(gwConfig.Link).Attrs().Index,
+			Scope:     netlink.SCOPE_LINK,
+			Dst:       c.nodeConfig.PodCIDR,
+			Table:     ServiceRtTable.Idx,
+		}
+		if err := netlink.RouteReplace(route); err != nil {
+			klog.Fatalf("Failed to add link route to service table: %v", err)
+		}
+	}
+
+	// create ip rule to select route table
+	ipRule := netlink.NewRule()
+	ipRule.IifName = util.GetNetLink(c.nodeConfig.GatewayConfig.Link).Attrs().Name
+	ipRule.Mark = iptables.RtTblSelectorValue
+	ipRule.Table = ServiceRtTable.Idx
+	ipRule.Priority = AntreaIPRulePriority
+
+	err = netlink.RuleAdd(ipRule)
+	if err != nil {
+		if err != unix.ENOENT {
+			klog.Errorf("rule already created")
+		} else {
+			klog.Fatalf("Failed to create ip rule for service route table: %v", err)
+		}
+	}
+	return nil
+}
+
+// AddPeerCIDRRoute adds routes to route tables for Antrea use.
+func (c *Client) AddPeerCIDRRoute(peerPodCIDR *net.IPNet, gwLinkIdx int, peerNodeIP, peerGwIP net.IP) ([]*netlink.Route, error) {
+	if peerPodCIDR == nil {
+		return nil, fmt.Errorf("empty peer pod CIDR")
+	}
+
+	// install routes
+	routes := []*netlink.Route{
+		{
+			Dst:       peerPodCIDR,
+			Flags:     int(netlink.FLAG_ONLINK),
+			LinkIndex: gwLinkIdx,
+			Gw:        peerGwIP,
+			Table:     ServiceRtTable.Idx,
+		},
+	}
+
+	// If service route table and main route table is not the same , add
+	// peer CIDR to main route table too (i.e in NoEncap and hybrid mode)
+	if !ServiceRtTable.IsMainTable() {
+		if c.encapMode.NeedsEncapToPeer(peerNodeIP, c.nodeConfig.NodeIPAddr) {
+			// need overlay tunnel
+			routes = append(routes, &netlink.Route{
+				Dst:       peerPodCIDR,
+				Flags:     int(netlink.FLAG_ONLINK),
+				LinkIndex: gwLinkIdx,
+				Gw:        peerGwIP,
+			})
+		} else if !c.encapMode.NeedsRoutingToPeer(peerNodeIP, c.nodeConfig.NodeIPAddr) {
+			routes = append(routes, &netlink.Route{
+				Dst:       peerPodCIDR,
+				LinkIndex: util.GetNetLink(c.nodeConfig.NodeDefaultDev).Attrs().Index,
```

Comment:
Do we must provide the dev in the route? I thought dev is optional?

Reply:
I am not sure, but do u see problem provide dev? It make clear to me where we want to remove the route from

Reply:
No, but if we could also avoid to look up the default interface/IP, but just use the node IP, it sounds better to me. I would not introduce another dependency, and would not discover host routes if possible.

Reply:
Revoke what I said since I found you need to look up the subnet anyway. But for this route, I feel better to let host stack decide which dev to use (for example there can be a more optimal route than using the Node IP dev)?

Reply:
line 164-169 is for peer node in hybrid or noEncap on same subnet as local node, in another word,  localIP, peerIP in same subnet, and localIP is assigned on default device. Thus if system find a better outbound interface (then directly connected route), either system messed up or there is bug is this Antrea route implementation.
what do u think?

Reply:
Maybe the other NICs perform better or user just wants to use it for Pod traffic? The IP can move from one NIC to other too and we have no way to detect? But my point is if it is not necessary why we worry about which dev to use?

Reply:
The question is, is NodeIP externally reachable  IP or not, i.e it is not 127.0.01.  The way Antrea and I suspect every other implementation assumes is that node IP is externally reachable IP. In that context, I think it is safe to say it does not matter which interface pod traffic goes out of for as long as interface is connected to the same physical l2 segment?

Reply:
sounds good, will change

Reply:
done